src/datasets/arrow_dataset.py

```diff
@@ -939,6 +939,45 @@ def from_csv(
             path_or_paths, split=split, features=features, cache_dir=cache_dir, keep_in_memory=keep_in_memory, **kwargs
         ).read()
 
+    @staticmethod
+    def from_generator(
+        generator: Callable,
+        features: Optional[Features] = None,
+        cache_dir: str = None,
+        keep_in_memory: bool = False,
+        gen_kwargs: Optional[dict] = None,
+    ):
+        """Create a Dataset from a generator.
+
+        Args:
+            generator (:obj:`Callable`): A callable object that returns an object that supports the `iter` protocol.
```

Comment:
(nit) suggestion for the docstring
```suggestion
            generator (:obj:`Callable`): A generator function that `yields` examples.
```